transportable-udfs-spark_2.11/src/main/scala/com/linkedin/transport/spark/StdUdfWrapper.scala

```diff
@@ -27,14 +27,17 @@ abstract class StdUdfWrapper(_expressions: Seq[Expression]) extends Expression
   @transient private var _stdFactory: StdFactory = _
   @transient private var _stdUdf: StdUDF = _
   @transient private var _requiredFilesProcessed: Boolean = false
-  @transient private var _outputDataType: DataType = _
+  @transient lazy private val _outputDataType: DataType = initialize()
   private var _nullableArguments: Array[Boolean] = _
   private var _distributedCacheFiles: Array[String] = _
 
   override def nullable: Boolean = true
 
-  override def dataType: DataType = {
-    if (_outputDataType != null) _outputDataType else initialize()
+  override def dataType: DataType = _outputDataType
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    _outputDataType
```

Comment:
How about making initialize do strictly initialize work:
```scala
  private def initialize(): DataType = {
    val sparkTypeInference = new SparkTypeInference
    sparkTypeInference.compile(children.map(_.dataType).toArray, getStdUdfImplementations, getTopLevelUdfClass)
    _stdFactory = sparkTypeInference.getStdFactory
    _stdUdf = sparkTypeInference.getStdUdf
    _nullableArguments = _stdUdf.getAndCheckNullableArguments
    _stdUdf.init(_stdFactory)
    getRequiredFiles()
    _requiredFilesProcessed = false
    sparkTypeInference.getOutputDataType
  }
```
and making `sparkTypeInference` a member variable that is used in `dataType`?

If needed to be called explicitly on the executor side, we can make that as part of `processRequiredFiles` workflow, probably after renaming the variable `_requiredFilesProcessed` to something more generic.

Reply:
In fact, if we ensure it is required on the executor side, I am not sure we need `checkInputDataTypes` at all. So in my opinion, we either need support in (`checkInputDataTypes` + `dataType`) or (`eval` + `dataType`), but not all three.

Reply:
I think making (`eval` + `dataType`) change should work, I will remove the change of `checkInputDataTypes` .